src/Collection/CollectionTrait.php

```diff
@@ -686,6 +686,47 @@ public function _unwrap()
      *
      * @return \Cake\Collection\CollectionInterface
      */
+    public function cartesianProduct(callable $operation = null, callable $filter = null)
+    {
+        if ($this->isEmpty()) {
+            return new Collection([]);
+        }
+
+        $result = [];
+        $counts = $this->map(function ($arr) {
+            return count($arr);
+        })->toList();
+        $allArr = $this->toList();
+        $lastIndex = count($allArr) - 1;
+        // holds the indexes of the arrays that generate the current combination
+        $currentIndexes = array_fill(0, $lastIndex + 1, 0);
```

Comment:
Yes. I preferred to call `count()` once. `$lastIndex` is used inside the loop.

So, have to choose between

- Call `count()` mutliple times

- Use `$lastIndex + 1` (as it is now)

- Use `$count = count(..)` and then use `$count - 1` where needed